engine/api/src/main/scala/pl/touk/nussknacker/engine/api/typed/CanBeSubclassDeterminer.scala

```diff
@@ -14,93 +17,153 @@ import pl.touk.nussknacker.engine.api.typed.typing._
   */
 trait CanBeSubclassDeterminer {
 
+  import CanBeSubclassDeterminerHelper._
+
   /**
     * java.math.BigDecimal is quite often returned as a wrapper for all kind of numbers (floating and without floating point).
     * Given to this we cannot to be sure if conversion is safe or not based on type (without scale knowledge).
     * So we have two options: enforce user to convert to some type without floating point (e.g. BigInteger) or be loose in this point.
     * Be default we will be loose.
     */
-    // TODO: Add feature flag: strictBigDecimalChecking (default false?) and rename strictTypeChecking to strictClassesTypeChecking
+  // TODO: Add feature flag: strictBigDecimalChecking (default false?) and rename strictTypeChecking to strictClassesTypeChecking
   private val ConversionFromClassesForDecimals = NumberTypesPromotionStrategy.DecimalNumbers.toSet + classOf[java.math.BigDecimal]
 
   /**
     * This method checks if `givenType` can by subclass of `superclassCandidate`
     * It will return true if `givenType` is equals to `superclassCandidate` or `givenType` "extends" `superclassCandidate`
     */
-  def canBeSubclassOf(givenType: TypingResult, superclassCandidate: TypingResult): Boolean =
+  def canBeSubclassOf(givenType: TypingResult, superclassCandidate: TypingResult): ValidatedNel[String, Unit] = {
     (givenType, superclassCandidate) match {
-      case (_, Unknown) => true
-      case (Unknown, _) => true
+      case (_, Unknown) => ().validNel
+      case (Unknown, _) => ().validNel
       case (given: SingleTypingResult, superclass: TypedUnion) => canBeSubclassOf(Set(given), superclass.possibleTypes)
       case (given: TypedUnion, superclass: SingleTypingResult) => canBeSubclassOf(given.possibleTypes, Set(superclass))
       case (given: SingleTypingResult, superclass: SingleTypingResult) => singleCanBeSubclassOf(given, superclass)
       case (given: TypedUnion, superclass: TypedUnion) => canBeSubclassOf(given.possibleTypes, superclass.possibleTypes)
     }
+  }
 
-  private def canBeSubclassOf(givenTypes: Set[SingleTypingResult], superclassCandidates: Set[SingleTypingResult]): Boolean =
-    givenTypes.exists(given => superclassCandidates.exists(singleCanBeSubclassOf(given, _)))
+  private def canBeSubclassOf(givenTypes: Set[SingleTypingResult], superclassCandidates: Set[SingleTypingResult]): ValidatedNel[String, Unit] = {
+    ensureOrInvalid(givenTypes.exists(given => superclassCandidates.exists(singleCanBeSubclassOf(given, _).isValid)))(
+      f"None of the following types:\n${givenTypes.map(" - " + _.display).mkString(",\n")}\ncan be a subclass of any of:\n${superclassCandidates.map(" - " + _.display).mkString(",\n")}"
```

Comment:
Why you user f interpolation everywhere? We generally use s interpolation where there is no formatting. Also For multiline message whould be better to user triple double quote (`""""`) with stripMargin

Reply:
I will refactor these things ;)